x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheServiceTests.java

```diff
@@ -0,0 +1,159 @@
+            cacheService.start();
+
+            logger.debug("--> setting large cache sync interval (explicit cache synchronization calls in test)");
+            cacheService.setCacheSyncInterval(TimeValue.timeValueMillis(Long.MAX_VALUE));
+
+            // Keep a count of the number of writes for every cache file existing in the cache
+            final Map<CacheKey, Tuple<CacheFile, Integer>> previous = new HashMap<>();
+
+            for (int iteration = 0; iteration < between(1, 10); iteration++) {
+
+                final Map<CacheKey, Tuple<CacheFile, Integer>> updates = new HashMap<>();
+
+                logger.trace("--> more random reads/writes from existing cache files");
+                for (Map.Entry<CacheKey, Tuple<CacheFile, Integer>> cacheEntry : randomSubsetOf(previous.entrySet())) {
+                    final CacheKey cacheKey = cacheEntry.getKey();
+                    final CacheFile cacheFile = cacheEntry.getValue().v1();
+
+                    final CacheFile.EvictionListener listener = evictedCacheFile -> {};
+                    cacheFile.acquire(listener);
+
+                    final SortedSet<Tuple<Long, Long>> newCacheRanges = randomPopulateAndReads(cacheFile);
+                    assertThat(cacheService.isCacheFileToSync(cacheFile), is(newCacheRanges.isEmpty() == false));
+                    if (newCacheRanges.isEmpty() == false) {
+                        final int numberOfWrites = cacheEntry.getValue().v2() + 1;
+                        updates.put(cacheKey, Tuple.tuple(cacheFile, numberOfWrites));
+                    }
+                    cacheFile.release(listener);
+                }
+
+                logger.trace("--> creating new cache files and randomly read/write them");
+                for (int i = 0; i < between(1, 25); i++) {
+                    final ShardId shardId = new ShardId(index, randomIntBetween(0, numShards - 1));
+                    final String fileName = String.format(Locale.ROOT, "file_%d_%d", iteration, i);
+                    final CacheKey cacheKey = new CacheKey(snapshotId, indexId, shardId, fileName);
+                    final CacheFile cacheFile = cacheService.get(cacheKey, randomIntBetween(1, 10_000), shardsCacheDirs[shardId.id()]);
+
+                    final CacheFile.EvictionListener listener = evictedCacheFile -> {};
+                    cacheFile.acquire(listener);
+
+                    final SortedSet<Tuple<Long, Long>> newRanges = randomPopulateAndReads(cacheFile);
+                    assertThat(cacheService.isCacheFileToSync(cacheFile), is(newRanges.isEmpty() == false));
+                    updates.put(cacheKey, Tuple.tuple(cacheFile, newRanges.isEmpty() ? 0 : 1));
+                    cacheFile.release(listener);
+                }
+
+                logger.trace("--> evicting random cache files");
+                for (CacheKey evictedCacheKey : randomSubsetOf(Sets.union(previous.keySet(), updates.keySet()))) {
```

Comment:
Would be good to verify that once we evicted, we do not fsync these `CacheFile`s anymore.

Reply:
Right - I added f69dde9e9e911a02fbf728a378e65da8b56cba7b